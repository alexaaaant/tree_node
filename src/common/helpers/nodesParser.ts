import { Node, NodeRaw, NodeState } from '../../types/nodes';

export const rebuildTree = (rawNodes: NodeRaw[]): NodeState => {
    const dict: Record<NodeRaw['id'], Node> = {};
    const rootNodes: Array<Node['id']> = [];

    rawNodes.forEach((rawNode) => {
        dict[rawNode.id] = { ...rawNode, selected: false };

        if (rawNode.parent !== undefined) {
            const parentNode = dict[rawNode.parent];
            
            const children = parentNode.children || [];
            children.push(rawNode.id);

            parentNode.children = children;
        } else {
            rootNodes.push(rawNode.id);
        }
    });
    return { rootNodes, nodes: dict };
};
