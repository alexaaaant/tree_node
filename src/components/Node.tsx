import { FC, memo } from 'react';
import { useDispatch } from '../common/hooks/useDispatch';
import { useSelector } from '../common/hooks/useSelector';
import { groupSelected, nodeSelected } from '../slices/nodesSlice';
import { Node } from '../types/nodes';

interface NodeProps {
  nodeId: Node['id'];
}

export const NodeComponent: FC<NodeProps> = memo(({ nodeId }) => {
    const node = useSelector((state) => state.nodes.nodes[nodeId], 
                             (prevNode, newNode) => prevNode.selected === newNode.selected);
    const dispatch = useDispatch();

    const selectNode = () => {
      dispatch(nodeSelected(nodeId));
    };

    const selectGroup = (event: React.MouseEvent<HTMLDivElement>) => {
      event.preventDefault(); 
      dispatch(groupSelected(nodeId));
    };

    return (
      <div className="node">
        <div
          className={`node__name ${node.selected ? 'node__name_selected' : ''}`}
          onClick={selectNode}
          onContextMenu={selectGroup}
        >
          {node.name}
        </div>
        {node.children ? (
          <div className="node_children">
            {node.children.map((childId) => (
              <NodeComponent key={childId} nodeId={childId} />
            ))}
          </div>
        ) : (
          <></>
        )}
      </div>
    );
  });

NodeComponent.displayName = 'NodeComponent';
