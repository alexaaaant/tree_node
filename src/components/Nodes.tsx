import { NodeComponent } from './Node';
import { useEffect } from 'react';
import { rebuildTree } from '../common/helpers/nodesParser';
import rawNodes from '../nodes';
import { useDispatch } from '../common/hooks/useDispatch';
import { nodesAdded } from '../slices/nodesSlice';
import { useSelector } from '../common/hooks/useSelector';
import '../styles/nodes.css';

export const NodesContainer = () => {
  const dispatch = useDispatch();
  const rootNodes = useSelector((state) => state.nodes.rootNodes);

  useEffect(() => {
    const nodesData = rebuildTree(rawNodes);
    dispatch(nodesAdded(nodesData));
  },        []);

  return (
    <div className="nodes_container">
      {rootNodes.map((nodeId) => (
        <NodeComponent key={nodeId} nodeId={nodeId} />
      ))}
    </div>
  );
};
