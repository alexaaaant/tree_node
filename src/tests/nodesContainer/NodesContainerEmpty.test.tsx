import { NodesContainer } from '../../components/Nodes';
import { render } from '../helpers/test-utils';

jest.mock('../../nodes', () => ({
    __esModule: true,
    default: [],
}));

describe('NodesContainer', () => {
    it('should render an empty tree, if nodesData is empty', () => {
      const { container } = render(<NodesContainer />);
      expect(container).toMatchSnapshot();
    });
});
