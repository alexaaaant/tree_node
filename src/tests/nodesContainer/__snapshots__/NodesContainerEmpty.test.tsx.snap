// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodesContainer should render an empty tree, if nodesData is empty 1`] = `
<div>
  <div
    class="nodes_container"
  />
</div>
`;
